In [ ]:
IMT 574

2.10 Assignment 2: Linear Regression and Gradient Descent
    
TIMOTHY MBUTHIA

In [39]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import xlrd
from sklearn.linear_model import LinearRegression

PROBLEM 1 [6 POINTS]
Download Airline Costs dataset.The dataset has the following attributes:

Airline, Length of flight (miles), Speed of plane (miles per hour),Daily flight time per plane (hours),Population served (1000s), Total operating cost (cents per revenue ton-mile),Revenue tons per aircraft mile,Ton-mile load factor (proportion),Available capacity (tons per mile), Total assets ($100,000s), Investments and special funds ($100,000s),Adjusted assets ($100,000s)

Part A [3 Points]

Use a linear regression model to predict the number of customers each airline serves from its length of the flight and daily flight time per plane.

In [23]:
# Read the contents of the file into a DataFrame
df_airline = pd.read_csv("Assignment 2—Problem 1_ airline_costs.csv", header=None)

# create list of airline dataset column headers 

columns = [ 'Airline', 'Length of flight (miles)', 'Speed of plane (miles per hour)','Daily flight time per plane (hours)','Population served (1000s)', 'Total operating cost (cents per revenue ton-mile)',
           'Revenue tons per aircraft mile','Ton-mile load factor (proportion)','Available capacity (tons per mile)', 'Total assets ($100,000s)', 'Investments and special funds ($100,000s)','Adjusted assets ($100,000s)']

# Set column headers
df_airline.columns = columns

# Now df has column headers
df_airline.head()

Airline  Length of flight (miles)  Speed of plane (miles per hour)  \
0  All American                        57                              133   
1      American                       270                              216   
2       Bonanza                       100                              140   
3       Braniff                       176                              182   
4       Capital                       142                              167   

   Daily flight time per plane (hours)  Population served (1000s)  \
0                                 6.10                      20200   
1                                 6.93                      56928   
2                                 4.45                        183   
3                                 6.60                      11869   
4                                 7.47                      41097   

   Total operating cost (cents per revenue ton-mile)  \
0                                              116.3   
1                                               43.0   
2                                              141.5   
3                                               50.6   
4                                               51.0   

   Revenue tons per aircraft mile  Ton-mile load factor (proportion)  \
0                            0.96                              0.400   
1                            3.98                              0.689   
2                            0.79                              0.358   
3                            2.57                              0.557   
4                            2.68                              0.510   

   Available capacity (tons per mile)  Total assets ($100,000s)  \
0                               2.400                     21.13   
1                               5.776                   1436.53   
2                               2.207                      6.65   
3                               4.614                    160.30   
4                               5.255                    195.02   

   Investments and special funds ($100,000s)  Adjusted assets ($100,000s)  
0                                       3.21                        17.92  
1                                     165.22                      1271.31  
2                                       0.01                         6.64  
3                                       5.81                       154.49  
4                                       6.06                       188.96

In [24]:
# Select relevant columns
X = df_airline[['Length of flight (miles)', 'Daily flight time per plane (hours)']]
y = df_airline['Population served (1000s)']

# Feature names
feature_names = X.columns.tolist()  # Get column names as list

# Build the linear regression model
model = LinearRegression()
model.fit(X, y)

# Coefficients
intercept = model.intercept_
coefficients = model.coef_

In [ ]:
Report your model (linear equation).

In [25]:
# Print the model equation
print("Linear Regression Model:")
print("Population served (1000s) = {:.2f} + {:.2f} * Length of flight (miles) + {:.2f} * Daily flight time per plane (hours)".format(intercept, coefficients[0], coefficients[1]))

Linear Regression Model:
Population served (1000s) = -7792.07 + 183.30 * Length of flight (miles) + -213.33 * Daily flight time per plane (hours)


What is your predicted value for the number of customers served for a flight that is 200 miles in length and has a daily flight time per plane of 7.2 hours? 

In [26]:
# Predict for a flight that is 200 miles in length and has a daily flight time per plane of 7.2 hours
length_of_flight = 200
daily_flight_time = 7.2
predicted_customers_served = model.predict([[length_of_flight, daily_flight_time]])[0]

# Print the predicted value
print("Predicted number of customers served:", predicted_customers_served)

Predicted number of customers served: 27331.03461681924


/Users/wanjirungugi/anaconda3/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Part B [3 Points]

Next, build another regression model to predict the total assets of an airline from the customer served by the airline.

In [20]:
# Select relevant columns
X1 = df_airline[['Population served (1000s)']]
y1 = df_airline['Total assets ($100,000s)']

# Build the linear regression model
model = LinearRegression()
model.fit(X1, y1)

# Coefficients
intercept2 = model.intercept_
coefficients2 = model.coef_


Once again, report your model.

In [21]:
# Print the model equation
print("Linear Regression Model:")
print("Total assets ($100,000s) = {:.2f} + {:.2f} * Population served (1000s)".format(intercept2, coefficients2[0]))

Linear Regression Model:
Total assets ($100,000s) = -98.51 + 0.02 * Population served (1000s)


What is your prediction for total assets for an airline, given they serve 20,300,000 customers? 

In [22]:
# Predict total assets for an airline that serves 20,300,000 customers
customers_served = 20300
predicted_total_assets = model.predict([[customers_served]])[0]

# Print the predicted value
print("Predicted total assets for an airline serving 20,300,000 customers: ${:.2f}".format(predicted_total_assets))

Predicted total assets for an airline serving 20,300,000 customers: $341.08


/Users/wanjirungugi/anaconda3/lib/python3.11/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


PROBLEM 2 [4 POINTS]
In this exercise, you are going to use the kangaroo’s nasal dimension data.Download kangaroo’s nasal dimension data.

INSTRUCTIONS
Use the gradient descent algorithm to predict the optimal intercept and gradient for this problem. Report your gradient values. [4 points]

In [36]:
# Read Excel file with specified encoding
df_kangaroo = pd.read_excel("Assignment 2—Problem 2_ kangaroo’s nasal dimension data.xls", engine="xlrd")
df_kangaroo.head()

*** No CODEPAGE record, no encoding_override: will use 'iso-8859-1'


X    Y
0  609  241
1  629  222
2  620  233
3  564  207
4  645  247

In [41]:
X = df_kangaroo['X']
y = df_kangaroo['Y']

# Data normalization function
def normalize_data(X):
    mean = np.mean(X)
    std_dev = np.std(X)
    return (X - mean) / std_dev

# Normalize the data
X_normalized = normalize_data(X)

# Gradient Descent with normalized data
optimal_intercept, optimal_gradient = gradient_descent(X_normalized, y)

# Report gradient values
print("Optimal Intercept:", optimal_intercept)
print("Optimal Gradient:", optimal_gradient)


Optimal Intercept: 23.235577743827818
Optimal Gradient: 2.506984829577155


The optimal intercept and gradient (or slope) values obtained from the gradient descent algorithm represent the parameters of the linear equation that best fits the given data.

Optimal Intercept value indicates the point where the regression line intersects the y-axis when all independent variables (X) are zero. In this case, the value of approximately 23.24 suggests that when the nasal dimension (X) is zero, the estimated value of Y (or the dependent variable) is around 23.24.

Optimal Gradient (Slope)represents the rate of change of the dependent variable (Y) with respect to the independent variable (X). Here, the gradient value of approximately 2.51 implies that for each unit increase in the nasal dimension (X), the estimated value of Y increases by approximately 2.51 units.

The nasal dimension (X) has a positive linear relationship with the dependent variable (Y), with an increase of approximately 2.51 units in Y for each unit increase in X, and when X is zero, the estimated value of Y is approximately 23.24.